### PR TITLE
Test multiple resolvers via matrix

### PR DESCRIPTION
**Pull request by stackptr**
Runs the `test` job against multiple Stack resolvers using a matrix strategy.

Motivated by #6, to exercise each case of the preprocessor directive that tests the version of `persistent`. LTS 18.0 pulls in the earlier version while the current 18.25 tests the later version.

Also, pins the version of `stack-action` and bumps it and `stack-cache-action` to latest version.

### CODE REVIEW

**Comment by stackptr** [comment]
Should `aws-xray-client-persistent` be bumped to version `0.1.0.3` or `0.1.1`? I'm not familiar with this versioning scheme.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -8,9 +8,21 @@ on:
 jobs:
   test:
     runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        stack-yaml:
+          - stack.yaml
+          - stack-lts-18.0.yaml
+          - stack-lts-18.25.yaml
```
Comment (pbrisbin) [inline .github/workflows/ci.yml]: While these exercise the persistent thing, they're both the same GHC version. We like to use the matrix to hit various GHCs -- which should also exercise the persistent thing.

```
GHC 8.10  => lts-18.26
GHC 8.8   => lts-16.31
GHC 8.6   => lts-14.27
GHC 8.4   => lts-12.26
GHC 8.2   => lts-11.22
GHC 8.0   => lts-9.21
GHC 7.10  => lts-6.35
GHC 7.8   => lts-2.22
```

I typically do 12, 14, and 16. But if the older ones aren't easy to get compiling, just 16 in addition to the default is OK with me. (I also don't typically duplicate the current `stack.yaml`, so no 18 necessary.)

File: stack-lts-18.0.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+resolver: lts-18.0
+
+packages:
+- aws-xray-client
+- aws-xray-client-wai
+- aws-xray-client-persistent
+
+ghc-options:
```
Comment (pbrisbin) [inline stack-lts-18.0.yaml]: I might not worry about this. Let our warning-preferences be `stack.yaml`-only to just cut down on the amount of copy/paste.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -8,9 +8,21 @@ on:
...
     steps:
       - uses: actions/checkout@v2
-      - uses: freckle/stack-cache-action@v1.0.1
-      - uses: freckle/stack-action@main
+      - uses: freckle/stack-cache-action@v2
+        with:
+          stack-yaml: ${{ matrix.stack-yaml }}
+      - uses: freckle/stack-action@v3.1.4
         with:
           weeder: false
```
Comment (pbrisbin) [inline .github/workflows/ci.yml]: You can remove this. The new version of the Action doesn't run weeder. You'll probably want to add an `hlint` Job too.

**Review by pbrisbin — CHANGES_REQUESTED** [review]
Thanks for digging into this! Just blocking on picking different actual resolvers, to maximize their value.

**Comment by pbrisbin** [comment]
For deciding the next version, you should follow [this guide](https://github.com/freckle/guides/blob/main/haskell-open-source.md#versioning).

**Comment by stackptr** [comment]
I updated the branch protection rules to remove the required `test` status check (which no longer runs with this change) and added the three checks above as required.

**Review by pbrisbin — APPROVED** [review]

**Pull request merged**